layer_metadata_restorer.go

```diff
@@ -53,24 +54,63 @@ func (la *DefaultLayerMetadataRestorer) restoreStoreTOML(appMeta platform.Layers
 	return nil
 }
 
-func (la *DefaultLayerMetadataRestorer) restoreLayerMetadata(appMeta platform.LayersMetadata, cacheMeta platform.CacheMetadata, buildpacks []buildpack.GroupBuildpack) error {
+type layerToSha struct {
+	layerToShaMap map[string]string
+}
+
+func initializeLayerToSha() layerToSha {
+	return layerToSha{make(map[string]string)}
+}
+
+type BuildpackLayersToSha struct {
+	buildpacksToLayersShaMap map[string]layerToSha
+}
+
+func initializeBuildpackLayersToSha() BuildpackLayersToSha {
+	return BuildpackLayersToSha{make(map[string]layerToSha)}
+}
+
+// if the layer exists for the buildpack ID, its SHA will be returned. Otherwise, an empty string will be returned.
+func (bls *BuildpackLayersToSha) getShaByBuildpackLayers(buildpackID, layerName string) string {
+	if layerToSha, buildpackExists := bls.buildpacksToLayersShaMap[buildpackID]; buildpackExists {
+		if sha, layerExists := layerToSha.layerToShaMap[layerName]; layerExists {
+			return sha
+		}
+	}
+	return ""
+}
+
+func (bls *BuildpackLayersToSha) addLayerToMap(buildpackID, layerName, sha string) {
+	_, ok := bls.buildpacksToLayersShaMap[buildpackID]
+	if !ok {
+		bls.buildpacksToLayersShaMap[buildpackID] = initializeLayerToSha()
+	}
+	bls.buildpacksToLayersShaMap[buildpackID].layerToShaMap[layerName] = sha
+}
+
+func (la *DefaultLayerMetadataRestorer) restoreLayerMetadata(appMeta platform.LayersMetadata, cacheMeta platform.CacheMetadata, buildpacks []buildpack.GroupBuildpack, writeShaToFile bool) (BuildpackLayersToSha, error) {
```

Comment:
Having `writeShaToFile` in `restoreLayerMetadata` and in `writeLayerMetadata` makes it so a consumer could actually get themselves into a weird state. Do you think we could hoist `writeShaToFile` to `NewLayerMetadataRestorer` so symmetry between the read and write exists?

Reply:
I like your idea, but I just wanted to make sure we're on the same page:

The write flow is:
```
Restorer::Restore
LayerMetadataRestorer::Restore (DefaultLayerMetadataRestorer)
DefaultLayerMetadataRestorer::restoreLayerMetadata
DefaultLayerMetadataRestorer::writeLayerMetadata
bpLayer::writeSha
```

The read flow is:
```
Restorer::Restore
bpLayer::read
```

Adding `writeShaToFile` to `NewLayerMetadataRestorer` is a good idea in order to reduce the number of variables to:
```
LayerMetadataRestorer::Restore (DefaultLayerMetadataRestorer)
DefaultLayerMetadataRestorer::restoreLayerMetadata
DefaultLayerMetadataRestorer::writeLayerMetadata
```

We'll still need to use this extra variable when calling `bpLayer::writeSha` and when deciding whether to call `bpLayer::read`.

Are you ok with this or did you mean something else?

Reply:
Yeah - that is what I had in mind

Reply:
Please see https://github.com/buildpacks/lifecycle/pull/624/commits/8a9af603ec9565b21ee5e43723c5c52f4263ebd1.
I'm going to mark this conversation as resolved but please feel free to reopen it if you think that I didn't address your feedback.
Thanks!